### ISSUE TITLE
Add option to no-unused-vars when using object spreads to omit properties

### ISSUE THREAD

**Issue by basicdays** [eslint/eslint#6563]
**What version of ESLint are you using?**
2.13.1

**What rule do you want to change?**
I'd like to propose a new option to the `no-unused-vars` rule that does not flag variables used in situations when using the Object Spread Properties when omitting properties while copying objects.

The ES proposal for the Stage 2 Object Rest/Spread Properties enhancement can be found here: https://github.com/sebmarkbage/ecmascript-rest-spread

**What code should be flagged as correct with this change?**
The following code has an object `data` that will eventually reach a point where I'd like to copy the object and omit specific properties. In this case I'm copying the properties into `copiedData` and omitting property `one`.

```js
let data = {one: 1, two: 2, three: 3};
//...

let {one, ...copiedData} = data;
```

**What happens when the rule is applied to this code now?**
It will specify one as an unused variable. While this is true, having an option to disregard unused variables when used with an omit pattern would be great to have. One area that I've used this pattern quite often is in React when passing properties from one component to the next and omitting properties along the way.

**Comment by platinumazure** [comment]
Hi @basicdays, it looks like you're [proposing a rule change](http://eslint.org/docs/developer-guide/contributing/rule-changes). Could you please augment your original post to answer the questions asked on the linked documentation? Thanks!

**Comment by eslintbot** [comment]
Hi @basicdays, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by basicdays** [comment]
@platinumazure Thanks for pointing that out, I've updated the initial post.

**Comment by basicdays** [comment]
Just wanted to also give a heads up that apparently this pattern is getting recommended in the React documentation as well: https://facebook.github.io/react/warnings/unknown-prop.html

**Comment by platinumazure** [comment]
Awesome, thanks for editing your post. Seems reasonable to me as long as it's been behind an option. Let's see what the team thinks.

**Comment by alberto** [comment]
Why don't you write it as:
```
let {one} = data;
```
instead?

**Comment by basicdays** [comment]
@alberto good question. The end goal of this is to actually get all the rest of the props from `data`, minus `one`. In this case, what I'd actually need is the inverse of what you have as an example.

**Comment by alberto** [comment]
Oh, sorry I didn't read carefully enough and misunderstood the issue.

**Comment by mysticatea** [comment]
@alberto To use `copiedData`.  He will use rest property syntax to copy with deleting a property.

Though I think too early to add an option for the stage 2 proposal.

**Comment by alberto** [comment]
Mmm, good point.

**Comment by basicdays** [comment]
I can certainly understand shelving the idea until the proposal rises to higher stages. For now I've just been doing something like the following example to tell the linter to not flag it:

https://github.com/jquense/react-formal/commit/0551ca0ed27b5bfae7fb16dd7eb64ef6c028825c

I'd imagine a lot of people are probably going to be using this pattern over time, assuming the proposal doesn't get derailed. :)

Is there a way to make it a plugin? I wouldn't mind finally looking at how eslint works if you all want to hold off integrating it into eslint proper.

**Comment by alberto** [comment]
Yes, you can copy and alter the rule as a plugin.

Closing for now.

**Comment by mysticatea** [comment]
I guess you can use `context.markVariableAsUsed` method for this.
This is similar to `react/jsx-uses-vars` rule.

**Comment by basicdays** [comment]
Sounds good, I'll look into it. Thanks for the feedback.

**Comment by mnpenner** [comment]
Do we have a 3rd-party solution to this yet? I'd like this feature too. I use object spread to omit vars all the time; it's very common in React.

**Comment by basicdays** [comment]
Wish I've had time to write something for this, but haven't been able to get into it yet. I haven't looked to see if anyone else have written something yet though.
